src/addon/managertracker.js

```diff
@@ -0,0 +1,101 @@
+/*globals*/
+/*jshint node:true*/
+/**
+ * This is keeps track of all addon-managers that create their own storages and in turn keeps track of the
+ * branch-monitors.
+ * @author pmeijer / https://github.com/pmeijer
+ */
+
+'use strict';
+
+var Q = require('q'),
+    AddOnManager = require('./addonmanager');
+
+function ManagerTracker(mainLogger, gmeConfig, options) {
+    var addOnManagers = {
+            //:projectId
+        },
+        logger = mainLogger.fork('ManagerTracker');
+
+    function connectedWorkerStart(webgmeToken, projectId, branchName, callback) {
+        var addOnManager;
+
+        logger.debug('connectedWorkerStart', projectId, branchName);
+        if (!projectId || !branchName) {
+            return Q.reject(new Error('Required parameters were not provided: ' + projectId + ', ' + branchName + '.'))
+                .nodeify(callback);
+        }
+
+        addOnManager = addOnManagers[projectId];
+
+        if (!addOnManager) {
+            logger.debug('No previous addOns handled for project [' + projectId + ']');
+            addOnManager = new AddOnManager(projectId, logger, gmeConfig, options);
+            addOnManagers[projectId] = addOnManager;
+            addOnManager.addEventListener('NO_MONITORS', function (/*addOnManager_*/) {
+                delete addOnManagers[projectId];
+                addOnManager.close()
+                    .fail(function (err) {
+                        logger.error('Error closing addOnManger', err);
+                    });
+            });
+        } else {
+            logger.debug('AddOns already being handled for project [' + projectId + ']');
+        }
+
+        return addOnManager.initialize(webgmeToken)
+            .then(function () {
+                return addOnManager.monitorBranch(branchName);
+            })
+            .then(function () {
+                return {
+                    managers: Object.keys(addOnManagers).length,
+                    branchMonitors: Object.keys(addOnManager.branchMonitors).length
+                };
+            })
+            .nodeify(callback);
+    }
+
+    function connectedWorkerStop(projectId, branchName, callback) {
+        if (!projectId || !branchName) {
+            return Q.reject(new Error('Required parameters were not provided: ' + projectId + ', ' + branchName + '.'))
+                .nodeify(callback);
+        } else if (addOnManagers[projectId]) {
+            return addOnManagers[projectId].unMonitorBranch(branchName)
+                .nodeify(callback);
+        } else {
+            logger.debug('Request stop for non existing addOnManger', projectId, branchName);
+            return Q.resolve()
+                .nodeify(callback);
+        }
+    }
+
+    this.connectedWorkerStart = connectedWorkerStart;
```

Comment:
is there a reason why these funcitons are first defined and then assigend to 'this'?